src/lib/consensus/proof_of_stake.ml

```diff
@@ -491,48 +491,78 @@ module Vrf = struct
+      Snarky_taylor.Exp.params ~base
+        ~field_size_in_bits:Snark_params.Tick.Field.size_in_bits
 
-        So that we don't have to do division we check
+    (* TODO: Make efficient *)
+    let bigint_of_uint64 = Fn.compose Bigint.of_string UInt64.to_string
 
-        vrf_output * total_currency <= c * my_stake * 2^256
+    (*  Check approximately if
+        vrf_output / (c * 2^256) <= 1 - base^(my_stake / total_currency)
     *)
     let is_satisfied ~my_stake ~total_stake vrf_output =
-      of_bit_fold_lsb (Random_oracle.Digest.fold_bits vrf_output)
-      * of_uint64_exn (Amount.to_uint64 total_stake)
-      <= shift_left
-           (c * of_uint64_exn (Balance.to_uint64 my_stake))
-           Random_oracle.Digest.length_in_bits
+      let input =
+        (* get first params.per_term_precision bits of top / bottom.
+
+           This is equal to
+
+           floor(2^params.per_term_precision * top / bottom) / 2^params.per_term_precision
+        *)
+        let k = params.per_term_precision in
+        let top = bigint_of_uint64 (Balance.to_uint64 my_stake) in
+        let bottom = bigint_of_uint64 (Amount.to_uint64 total_stake) in
+        Bignum.(
+          of_bigint Bignum_bigint.(shift_left top k / bottom)
+          / of_bigint Bignum_bigint.(shift_left one k))
+      in
+      let rhs = Snarky_taylor.Exp.Unchecked.one_minus_exp params input in
+      let lhs =
+        let bs = Random_oracle.Digest.to_bits vrf_output in
+        of_bits_lsb (match c with `One -> bs | `Two -> List.tl_exn bs)
+      in
+      Bignum.(of_bigint lhs <= rhs)
 
     module Checked = struct
-      (* This version can't be used right now because the field is too small. *)
-      let _is_satisfied ~my_stake ~total_stake vrf_output =
-        let open Snark_params.Tick in
-        let open Let_syntax in
-        let open Number in
-        let%bind lhs = of_bits vrf_output * Amount.var_to_number total_stake in
-        let%bind rhs =
-          let%bind x =
-            (* someday: This should really just be a scalar multiply... *)
-            constant (Field.of_int c_int) * Amount.var_to_number my_stake
-          in
-          mul_pow_2 x (`Two_to_the Random_oracle.Digest.length_in_bits)
-        in
-        lhs <= rhs
-
-      (* It was somewhat involved to implement that check with the small field, so
-         we've stubbed it out for now. *)
-      let is_satisfied ~my_stake:_ ~total_stake:_ _vrf_output =
-        let () = assert Coda_base.Insecure.vrf_threshold_check in
-        Snark_params.Tick.(Checked.return Boolean.true_)
+      module M = Snarky.Snark.Run.Make (Snark_params.Tick_backend)
+
+      let m : M.field Snarky.Snark.m = (module M)
+
+      let is_satisfied ~my_stake ~total_stake (vrf_output : Output.var) =
+        (* TODO: Ask matthew about this. was thinking there would be a 
+           val direct : (field m, _) Checked.t *)
+        let open Snarky_taylor in
+        let open M in
+        make_checked (fun () ->
```

Comment:
I'm a little out of date on snarky development. Is this the new imperative interface?